### PR TITLE
Support for calculating the period associated with each node in a graph

### PR DESCRIPTION
**Pull request by ZacCranko**
Exports one new function 
* `periods(g::DiGraph)` that functions as advertised. Additionally there's a new internal function 
* `strongly_connected_period(g::DiGraph)`. The nodes in a strongly connected component all have the same period, so this returns the common period associated with all of the nodes in a strongly connected graph `g`.

Also:
* now we warn on self loop edges being added
* I don't think my tests for `strongly_connected_components` are great, could someone please double check this?

### CODE REVIEW

File: src/connectivity.jl [hunk 1]
```diff
@@ -83,3 +83,32 @@ function strongly_connected_components(g::DiGraph)
     end
     return components
 end
+
+# Computes the (common) period for all nodes in a strongly connected graph
+function strongly_connected_period(g::DiGraph)
```
Comment (sbromberger) [inline src/connectivity.jl]: Are you not exporting `strongly_connected_periods`?

File: test/graphdigraph.jl [hunk 2]
```diff
@@ -22,10 +22,10 @@ add_edge!(h, 2, 5)
 add_edge!(h, 3, 5)
 
 
-@test_throws ErrorException add_edge!(g, 1, 2)
-@test_throws ErrorException add_edge!(h, 1, 2)
-@test_throws ErrorException add_edge!(g, 1, 1)
-@test_throws ErrorException add_edge!(h, 1, 1)
+# @test_throws ErrorException add_edge!(g, 1, 2)
+# @test_throws ErrorException add_edge!(h, 1, 2)
```
Comment (sbromberger) [inline test/graphdigraph.jl]: Lines 25 and 26 should still remain, no? They're not self-loops - they're duplicate edges.

**Comment by sbromberger** [comment]
See line comments. Also - is it possible to squash the commits a bit?

File: src/connectivity.jl (hunk 1 above)
Comment (ZacCranko) [inline src/connectivity.jl]: In short I'm undecided. My reason for not exporting this is that `strongly_connected_periods` returns a single integer, and `periods` returns a list of int's, one for each node. I'm happy to export both incase the user knows whether the graph will be strongly connected ahead of time.

**Comment by sbromberger** [comment]
Also - we need docs. If you can type something up I'll format it and put it in the proper place.

File: test/graphdigraph.jl (hunk 2 above)
Comment (ZacCranko) [inline test/graphdigraph.jl]: Whoops, I'll fix this now.

**Comment by ZacCranko** [comment]
Hi @oyamad, the place to discuss this PR is here. Several things are going to change, at any rate I'll be attending to this tomorrow morning.

**Comment by oyamad** [comment]
* Do you have a plan to add to `period` the functionality to return cyclic components as well, which can be obtained simultaneously with the period?

* In `period`, you might check if the digraph has a self loop in the beginning; if it does, the period is immediately determined to be 1 without performing BFS.

**Comment by sbromberger** [comment]
Self loop detection should be easy and a straightforward implementation would be O(V) worst case:

`has_self_loop(g:AbstractGraph) = any(v->has_edge(g, v, v), vertices(g))`

**Comment by ZacCranko** [comment]
Great work guys, this is much better now.

**Comment by sbromberger** [comment]
It'd be great if you added the has_self_loops as a function in core.jl and used it. I can see it being useful in other places too.

**Comment by ZacCranko** [comment]
You read my mind seth

**Comment by sbromberger** [comment]
Update: bfsdfs is working but here is still a bug in Compat.jl that's causing a less-than-optimal code issue (see issue #s 110 and 111 in Compat.jl if you're interested in the details). As soon as that's resolved I'll merge it and then look to merge this as well. It'd be great if you had documentation for the new functionality that I could include.

**Comment by ZacCranko** [comment]
the commit chain for this is too mangled

**Pull request closed without merging**